module/localisation/RobotParticleLocalisation/src/RobotParticleLocalisation.cpp

```diff
@@ -64,48 +73,111 @@ namespace module::localisation {
             "Measurement Update",
             [this](const VisionGoals& goals, const FieldDescription& fd) {
                 if (!goals.goals.empty()) {
-                    /* Perform time update */
+                    // Perform time update
                     using namespace std::chrono;
                     const auto curr_time  = NUClear::clock::now();
                     const double seconds  = duration_cast<duration<double>>(curr_time - last_time_update_time).count();
                     last_time_update_time = curr_time;
 
                     filter.time(seconds);
 
-                    for (auto goal : goals.goals) {
-
-                        // Check side and team
-                        const Eigen::Vector3d poss = getFieldPosition(goal, fd);
-
-                        for (auto& m : goal.measurements) {
-                            if (m.type == VisionGoal::MeasurementType::CENTRE) {
-                                if (m.position.allFinite() && m.covariance.allFinite()) {
-                                    filter.measure(Eigen::Vector3d(m.position.cast<double>()),
-                                                   Eigen::Matrix3d(m.covariance.cast<double>()),
-                                                   poss,
-                                                   goals.Hcw,
-                                                   m.type,
-                                                   fd);
-                                }
-                                else {
-                                    log("Received non-finite measurements from vision. Discarding ...");
+                    // Currently this has no idea about pairs of posts, only that it has been labelled left, right and
+                    // unknown
+
+                    // Save the filter state
+                    auto saved_filter = filter;
+
+                    // Go through all of the known sided posts
+                    for (auto goal_post : goals.goals) {
+                        if (goal_post.side != Goal::Side::UNKNOWN_SIDE) {
+                            // These are either left or right goal posts
+
+                            // Compare each of these to the possible goal posts on the field (own and opp)
+
+                            // TODO remove the repeated measurements
+                            auto m = goal_post.measurements[0];
+
+                            // Check that the measurement is sane
+                            if (m.position.allFinite() && m.covariance.allFinite()) {
+                                auto filter_new_own = saved_filter;
+
+                                // Make a new candidate for the post
+                                auto candidate_own =
```

Comment:
Maybe give a better name to these, candidate isn't really clear. They really are just likelihoods this filter thinks it's real
```suggestion
                                // Run a measurement for the post and store how likely the filter thinks this is a real measurement
                                auto own_logits =
```